src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -504,15 +548,36 @@ bool ShenandoahOldHeuristics::should_start_gc() {
 
   if (_fragmentation_trigger) {
     ShenandoahHeap* heap = ShenandoahHeap::heap();
+    const size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
+
     size_t used = _old_generation->used();
     size_t used_regions_size = _old_generation->used_regions_size();
+
+    // used_regions includes humongous regions
     size_t used_regions = _old_generation->used_regions();
     assert(used_regions_size > used_regions, "Cannot have more used than used regions");
+
+    size_t first_active_region = heap->free_set()->first_old_region();
+    size_t last_active_region = heap->free_set()->last_old_region();
+    size_t span_of_active_regions = (last_active_region > first_active_region)? last_active_region + 1 - first_active_region: 0;
+
+    size_t first_old_region, last_old_region;
+    double density;
+    fragmentation_trigger_reason(density, first_old_region, last_old_region);
+    size_t span_of_old_regions = (last_old_region > first_old_region)? last_old_region + 1 - first_old_region: 0;
+
     size_t fragmented_free = used_regions_size - used;
-    double percent = percent_of(fragmented_free, used_regions_size);
+
+    // New active regions may have came into play following the trigger.
+    size_t first_region = MIN2(first_active_region, first_old_region);
```

Comment:
I agree this is confusing.  I'm going to simplify this code and have the log message report the data that drove the trigger.  The reality is that between the moment in time that we raised the trigger (from inside ShenandoahHeap::rebuild_free_set(), and the moment that the regulator polls the old heuristic's state to discover that we should_start_gc(), the state may change so this log message may be slightly out of date.  This would be extremely unlikely.  The trigger happens at final-update-refs.  The old-gen state will not change until we begin a subsequent evacuation phase.

Reply:
Will simplify this code to make it more clear.